### DISCUSSION TITLE
Running autoresearch on Jetson: ARM64 port + automated workflow guards

### DISCUSSION THREAD

**Discussion by radozaprazny** [karpathy/autoresearch#499] [Show and tell]
# Jetson port + workflow integrity checks that made 102 experiments fully reproducible (3.50% improvement)

## 🎯 The Problem

While porting autoresearch to Jetson, I kept hitting the same workflow failures:

- 🔴 Running experiments with uncommitted changes → impossible to reproduce which commit produced result
- 🔴 OOM crashes from oversized models → hours wasted on 8GB unified memory
- 🔴 Accidentally editing code mid-experiment → corrupted results with no trace

**After 10+ failed runs, I added workflow guards to automate these checks.**

## ✅ The Solution: Automated Workflow Improvements

Added 3 code-enforced checks and 1 agent guideline (~30 lines of code total, platform-agnostic):

### 1. **Git Dirty Check** — Reproducibility Guard
```python
# train.py:458-465
result = subprocess.run(['git', 'diff-index', '--quiet', 'HEAD'], capture_output=True)
if result.returncode != 0:
    print("❌ ERROR: Uncommitted changes detected")
    print("   WORKFLOW: git commit → python train.py")
    sys.exit(1)
```
**Impact:** Every experiment has matching git commit. Zero ambiguity. Script refuses to run on uncommitted changes.

### 2. **VRAM Pre-flight Check** — OOM Prevention
```python
# train.py:504-507
_vram_est_gb = num_params * 2 / 1e9 * 3.5
if _vram_est_gb > 6.8:
    print(f"⚠️  WARNING: VRAM Est ~{_vram_est_gb:.1f}GB -> Potential OOM")
```
**Impact:** Catches oversized models before training starts. Zero OOM crashes across 102 experiments on 8GB Jetson.

### 3. **Script Hash Tracking** — Audit Trail
```python
# train.py:651-652
with open(__file__, 'rb') as f:
    _script_hash = hashlib.sha256(f.read()).hexdigest()[:8]
print(f"script_hash:      {_script_hash}")
```
**Impact:** Provides an audit trail for every experiment output. Note: post-commit edits are already blocked by the git dirty check — script hash adds a second layer of verification if the check is ever bypassed.

### 4. **Anti-Stagnation Rule** — Encourage Structural Exploration
After 5 consecutive discards with <0.002 improvement (program.md):
- ❌ **Discouraged:** LR tweaks, weight decay micro-adjustments
- ✅ **Encouraged:** Activation functions, attention mechanisms, architectural changes

**Impact:** Guideline for the LLM agent (not code-enforced). Pushed early exploration toward structural changes — the most impactful improvements came in experiments 1–65. The rule didn't fully prevent late-run micro-tuning (~20 final experiments), but the structural phase produced all meaningful gains.

## 📊 Results: Jetson Orin Nano Super (ARM64)

**Hardware:** 8GB unified memory, 1024 CUDA cores, ARM64, JetPack 6.1  
**Model:** 1.6 GB (27× smaller than H100's 44GB model)  
**Steps:** ~1350 per 5-min experiment (vs ~8000 on H100)

- ✅ Total experiments: 102
- ✅ Baseline: 1.454936 bpb
- ✅ Best: 1.404085 bpb (commit `2e6bd5b`)
- ✅ Improvement: **3.50%**
- ✅ Keep rate: 19.6% (20/102)
- ✅ **100% git compliance** — every experiment has clean git state
- ✅ **Zero crashes** — 0 OOM errors, 0 config violations, all experiments completed

## 🔧 Code Changes

**Platform (ARM64 compatibility):**
- Flash Attention 3 → PyTorch SDPA
- Disabled `torch.compile` (no Triton on ARM64)
- Reduced model depth 8→4, aspect ratio 96→64 (apr4) / 60 (master)

**Workflow improvements (platform-independent):**
- `train.py:458-465` — git dirty check (enforces commit-before-run)
- `train.py:504-507` — VRAM pre-flight check (prevent OOM)
- `train.py:651-652` — script hash logging (audit trail)
- `program.md:36-49` — anti-stagnation guideline (encourage structural exploration)

## 📦 Repo

**Fork:** [radozaprazny/autoresearch-jetson](https://github.com/radozaprazny/autoresearch-jetson)

**Two branches to show incremental value:**

| Branch | Purpose | Files |
|--------|---------|-------|
| `master` | Baseline Jetson port (ARM64 compatibility only) | Only platform changes |
| `autoresearch/apr4` ⭐ | **With integrity checks** (recommended) | + [results.tsv](https://github.com/radozaprazny/autoresearch-jetson/blob/autoresearch/apr4/results.tsv) (102 experiments) |

**Progress graph:**

![Jetson apr4 progress](https://github.com/radozaprazny/autoresearch-jetson/raw/master/progress.png?raw=true)

*Shows validation loss (bpb) over 102 experiments. Green points = kept improvements, gray = discarded. Annotations show commit messages.*

> 💡 Check out `autoresearch/apr4` branch to see complete results and workflow implementation.

---

**Comment by GreyforgeLabs** [comment]
This is one of the more interesting follow-ons here because the workflow guards matter more than the hardware port.

Dirty-tree refusal, resource preflight, and script-hash logging are all signs that integrity constraints start to matter once the loop gets used seriously. From the Greyforge side, that is closer to the real frontier than just proving the same benchmark loop runs on one more device.

It is still local and lightweight, but it points in the right direction: stronger provenance, stronger auditability, and fewer ways for the loop to quietly lie to you.
